test/generate/build_test.dart

```diff
@@ -0,0 +1,224 @@
+          ]),
+        ],
+        [
+          new Phase([
+            new CopyBuilder(numCopies: 2, outputPackage: 'a')
+          ], [
+            new InputSet('a', filePatterns: ['lib/*.txt.*']),
+            new InputSet('b', filePatterns: ['**/*.dart']),
+          ]),
+        ]
+      ];
+      await testPhases(phases, {
+        'a|web/1.txt': '1',
+        'a|lib/2.txt': '2',
+        'b|lib/3.txt': '3',
+        'b|lib/b.dart': 'main() {}',
+      }, outputs: {
+        'a|web/1.txt.copy': '1',
+        'a|lib/2.txt.copy': '2',
+        'a|lib/3.txt.clone': '3',
+        'a|lib/2.txt.copy.copy.0': '2',
+        'a|lib/2.txt.copy.copy.1': '2',
+        'a|lib/3.txt.clone.copy.0': '3',
+        'a|lib/3.txt.clone.copy.1': '3',
+        'a|lib/b.dart.copy.0': 'main() {}',
+        'a|lib/b.dart.copy.1': 'main() {}',
+      });
+    });
+  });
+
+  group('errors', () {
+    test('when overwriting files', () async {
+      var phases = [
+        [
+          new Phase([new CopyBuilder()], [new InputSet('a')]),
+        ]
+      ];
+      await testPhases(phases, {'a|lib/a.txt': 'a', 'a|lib/a.txt.copy': 'a'},
+          status: BuildStatus.Failure,
+          exceptionMatcher: invalidOutputException);
+    },
+        skip: 'InMemoryAssetWriter doesn\'t check this, should be handled at '
+            'the BuildStep level.');
+  });
+
+  test('outputs from builds shouldn\'t be inputs to later ones', () {},
+      skip: 'Unimplemented');
+}
+
+testPhases(List<List<Phase>> phases, Map<String, String> inputs,
+    {Map<String, String> outputs,
+    PackageGraph packageGraph,
+    BuildStatus status: BuildStatus.Success,
+    exceptionMatcher}) async {
+  final writer = new InMemoryAssetWriter();
+  final actualAssets = writer.assets;
+  final reader = new InMemoryAssetReader(actualAssets);
+  inputs.forEach((serializedId, contents) {
+    writer.writeAsString(makeAsset(serializedId, contents));
+  });
+
+  if (packageGraph == null) {
+    var rootPackage = new PackageNode('a', null, null, null);
+    packageGraph = new PackageGraph.fromRoot(rootPackage);
+  }
+
+  var result = await build(phases,
+      reader: reader, writer: writer, packageGraph: packageGraph);
+  expect(result.status, status);
+  if (exceptionMatcher != null) {
+    expect(result.exception, exceptionMatcher);
+  }
+
+  if (outputs != null) {
+    var remainingOutputIds =
+        new List.from(result.outputs.map((asset) => asset.id));
+    outputs.forEach((serializedId, contents) {
+      var asset = makeAsset(serializedId, contents);
+      remainingOutputIds.remove(asset.id);
+
+      /// Check that the writer wrote the assets
+      expect(actualAssets, contains(asset.id));
+      expect(actualAssets[asset.id], asset.stringContents);
+
+      /// Check that the assets exist in [result.outputs].
+      var actual = result.outputs
+          .firstWhere((output) => output.id == asset.id, orElse: () => null);
+      expect(actual, isNotNull);
```

Comment:
it might be worth adding a reason here and below to make it easier to recognize why a test is failing.